### ISSUE TITLE
Roadmap for Functional Components?

### ISSUE THREAD

**Issue by chadwtkns** [Crownpeak/DXM-React-SDK#7]
Hi we are currently utilizing functional components for our applications and havent really introduced the idea of class components yet. We are using crownpeak in other parts of our company and are exploring bringing it to this application. Are there plans to have a version of the SDK that supports a functional component paradigm? Thank you!

**Comment by richard-lund** [comment]
Currently we haven't got any plans to support functional components within the React SDK. The approach that we took when writing the SDK is to parse out components from the source code and use those to make updates to the CMS. It would be a significant challenge to recognise a component when the only characteristic that makes it that way is the fact that it returns a React object.

With that said, the code is released under an MIT license, so if you want to make enhancements such as this, you're very welcome to fork the repo and help yourself. If you do, please consider sending in a pull request with your updates so that the wider community can benefit from your improvements!

**Comment by chadwtkns** [comment]
Would you guys be open to the idea of having a component be recognized based off of a naming convention? Something like ending component name with 'CPComponent' or something along those lines. Im just thinking out loud here so definitely poke holes or if that just not a direction your interested in.

**Comment by richard-lund** [comment]
Reopening this to canvas for votes. If anyone else is interested in using the DXM React SDK with functional components rather than class-based ones, please up-vote to help us to gauge the level of interest.

We will carry out a level-of-effort estimate once a functional application sample is available.

**Comment by richard-lund** [comment]
@chadwtkns - I have a first draft of this working in React. Does this look like the application that you already have?

CmsComponent:
```
function HeroContainer(props)
{
    CmsDataCache.setComponent("HeroContainer");

    var heading = new CmsField("Heading", CmsFieldTypes.TEXT, props && props.data ? props.data.Heading : null, CmsIndexedField.STRING);

    return (
        <div className="container">
            <h1 className="display-3">{ heading }</h1>
        </div>
    )
}
```

CmsPage:
```
function HomePage(props) {
    CmsStaticPage.load(266812);

    return (
            <div className="jumbotron">
                <HeroContainer />
            </div>
    )
}
```

If this looks like what you have, I can figure out how to have the parsers detect this format.

**Comment by chadwtkns** [comment]
Hey Richard is there a branch of the SDK to test this from? When I try to utilize the master branch it says that 'load' is not a function and I cant find a reference to a load function in the SDK. I could be totally blind though

**Comment by chadwtkns** [comment]
I do think the idea should work though. I was going about in a more complex way of trying to build pages and components as hooks

**Comment by richard-lund** [comment]
It's a bit tricky because it needs changes to two packages, so I didn't make a branch at this point. I'll email you the changed files and some instructions and you can try it out.

**Comment by richard-lund** [comment]
It's a bit tricky because it needs changes to two packages, so I didn't make a branch at this point. I have emailed you the changed files and some instructions for you to try it out.

**Comment by richard-lund** [comment]
@chadwtkns it sounds like your email didn't allow my zip file through, so I'll attach here:

[sdk-updates.zip](https://github.com/Crownpeak/DXM-React-SDK/files/5369819/sdk-updates.zip)

Instructions again:
1. Make sure you have version 2.4.0 of the crownpeak-dxm-react-sdk package.
2. Copy the files in the attached zip into the relevant places inside the node_modules folder.
3. Copy in the component code from the example above.
4. Give it some JSON in the shape your component expects.

**Comment by richard-lund** [comment]
Support for function components was released with version 3.0.0.
